File: mobile/ios/packages/town-crier-data/Sources/Repositories/APIZonePreferencesRepository.swift
import Foundation
import TownCrierDomain

/// Fetches and updates per-zone notification preferences via the Town Crier API.
public final class APIZonePreferencesRepository: ZonePreferencesRepository, Sendable {
  private let apiClient: URLSessionAPIClient

  public init(apiClient: URLSessionAPIClient) {
    self.apiClient = apiClient
  }

  public func fetchPreferences(zoneId: String) async throws -> ZoneNotificationPreferences {
    let dto: ZonePreferencesDTO
    do {
      dto = try await apiClient.request(.get("/v1/me/watch-zones/\(zoneId)/preferences"))
    } catch let domainError as DomainError {
      throw domainError
    } catch {
      throw error.toDomainError()
    }
    return dto.toDomain()
  }

  public func updatePreferences(_ preferences: ZoneNotificationPreferences) async throws {
    let body = ZonePreferencesDTO(
      zoneId: preferences.zoneId,
      newApplicationPush: preferences.newApplicationPush,
      newApplicationEmail: preferences.newApplicationEmail,
      decisionPush: preferences.decisionPush,
      decisionEmail: preferences.decisionEmail
    )
    do {
      let _: EmptyResponse = try await apiClient.request(
        .put("/v1/me/watch-zones/\(preferences.zoneId)/preferences", body: body)
      )
    } catch let domainError as DomainError {
      throw domainError
    } catch {
      throw error.toDomainError()
    }
  }
}

// MARK: - DTO

struct ZonePreferencesDTO: Codable, Sendable {
  let zoneId: String
  let newApplicationPush: Bool
  let newApplicationEmail: Bool
  let decisionPush: Bool
  let decisionEmail: Bool

  func toDomain() -> ZoneNotificationPreferences {
    ZoneNotificationPreferences(
      zoneId: zoneId,
      newApplicationPush: newApplicationPush,
      newApplicationEmail: newApplicationEmail,
      decisionPush: decisionPush,
      decisionEmail: decisionEmail
    )
  }
}

File: mobile/android/data/src/main/kotlin/uk/towncrierapp/data/watchzones/ApiZonePreferencesRepository.kt
package uk.towncrierapp.data.watchzones

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import uk.towncrierapp.data.api.ApiClient
import uk.towncrierapp.data.api.ApiEndpoint
import uk.towncrierapp.domain.watchzones.WatchZoneId
import uk.towncrierapp.domain.watchzones.ZoneNotificationPreferences
import uk.towncrierapp.domain.watchzones.ZonePreferencesRepository

/** `ZonePreferencesRepository` over the Town Crier API. Port of iOS `APIZonePreferencesRepository`. */
public class ApiZonePreferencesRepository(
    private val apiClient: ApiClient,
    private val json: Json = Json { ignoreUnknownKeys = true },
) : ZonePreferencesRepository {
    override suspend fun fetchPreferences(zoneId: WatchZoneId): ZoneNotificationPreferences =
        apiClient
            .request(ApiEndpoint.get("/v1/me/watch-zones/${zoneId.value}/preferences"), ZonePreferencesDto.serializer())
            .toDomain()

    override suspend fun updatePreferences(preferences: ZoneNotificationPreferences) {
        val body = json.encodeToString(ZonePreferencesDto.serializer(), preferences.toDto())
        apiClient.requestBytes(
            ApiEndpoint.put("/v1/me/watch-zones/${preferences.zoneId.value}/preferences", body = body),
        )
    }
}

@Serializable
internal data class ZonePreferencesDto(
    val zoneId: String,
    val newApplicationPush: Boolean = true,
    val newApplicationEmail: Boolean = true,
    val decisionPush: Boolean = true,
    val decisionEmail: Boolean = true,
)

internal fun ZonePreferencesDto.toDomain(): ZoneNotificationPreferences =
    ZoneNotificationPreferences(
        zoneId = WatchZoneId(zoneId),
        newApplicationPush = newApplicationPush,
        newApplicationEmail = newApplicationEmail,
        decisionPush = decisionPush,
        decisionEmail = decisionEmail,
    )

internal fun ZoneNotificationPreferences.toDto(): ZonePreferencesDto =
    ZonePreferencesDto(
        zoneId = zoneId.value,
        newApplicationPush = newApplicationPush,
        newApplicationEmail = newApplicationEmail,
        decisionPush = decisionPush,
        decisionEmail = decisionEmail,
    )
